tests/operations/PatchOperationListTest.php

```diff
@@ -0,0 +1,369 @@
+                '[{"op":"add", "path": "/foo"}]',
+                InvalidPatchOperationException::class,
+                // The message also contains the original PHP message but this varies between versions so we do not
+                // include it in the assertion.
+                sprintf(
+                    'Missing required param(s) for add operation as %s',
+                    Add::class,
+                ),
+            ],
+            'operation property has incorrect type' => [
+                '[{"op":"add", "path": true, "value": "bar"}]',
+                InvalidPatchOperationException::class,
+                // The message also contains the original PHP message but this varies between versions so we do not
+                // include it in the assertion.
+                sprintf(
+                    'Invalid param(s) for add operation as %s',
+                    Add::class,
+                ),
+            ],
+            'operation with mixed string and int keys (example 1)' => [
+                // PHP would internally convert this into positional args and accept it even though there's no guarantee
+                // we have the right keys in the right sequence / have no extra properties.
+                '[{"op":"add", "0": "bar", "1": "/from"}]',
+                InvalidPatchOperationException::class,
+                'All patch operation properties must have string names',
+            ],
+            'operation with mixed string and int keys (example 2)' => [
+                // PHP would internally convert this into positional args and accept it even though there's no guarantee
+                // we have the right keys in the right sequence / have no extra properties.
+                '[{"op":"add", "1": "bar", "2": "/from"}]',
+                InvalidPatchOperationException::class,
+                'All patch operation properties must have string names',
+            ],
+        ];
+    }
+
+    /**
+     * @param string $json
+     * @param class-string<Throwable> $expect_exception
+     * @param string $expect_msg
+     * @return void
+     */
+    #[DataProvider('invalidJsonProvider')]
+    public function testItThrowsOnAttemptToCreateFromInvalidJson(string $json, string $expect_exception, string $expect_msg): void
+    {
+        $this->expectException($expect_exception);
+        $this->expectExceptionMessage($expect_msg);
+        PatchOperationList::fromJson($json);
+    }
+
+    public function testItsFromJsonBubblesGenericPhpErrors(): void
+    {
+        // Very contrived example to prove that our code rethrows any unexpected PHP errors during DTO creation
+        $this->expectException(DivisionByZeroError::class);
+        PatchOperationList::fromJson(
+            <<<'JSON'
+            [{"op": "fraction", "numerator": 1, "denominator": 0}]
+            JSON,
+            customClasses: ['fraction' => Fraction::class]
+        );
+    }
+}
+
+readonly class Append extends PatchOperation
```

Comment:
Unfortunately anonymous classes cannot be `readonly` until PHP 8.3, so the custom classes used by these tests have to be defined as actual named classes.